### ISSUE TITLE
Option for rewrap on resize

### ISSUE THREAD

**Issue by egmontkob** [mate-desktop/mate-terminal#101]
gnome-terminal has this, as per https://bugzilla.gnome.org/show_bug.cgi?id=336238

Rationale is that it synchronously rewraps the entire scrollback, and with giant scrollback it might be very slow (for me it's about 0.2 seconds for 1M lines of scrollback).

(Rewrap on resize is available since vte 0.36.)

**Comment by monsta** [comment]
> Rationale is that it synchronously rewraps the entire scrollback, and with giant scrollback it might be very slow (for me it's about 0.2 seconds for 1M lines of scrollback).

I'm somewhat confused by this description... is this the current behavior of gnome-terminal?

**Comment by egmontkob** [comment]
Yes. It's a synchronous (blocking) operation inside vte, and in turn it blocks gnome-terminal for its duration. (Just give it a try, you might have a somewhat faster computer than me but it won't be magnitudes faster. Set scrollback to 10M lines, produce that many lines and then resize.)

Also note that gnome-terminal resizes each tab only when you switch to that tab. I'm not sure how it goes with mate-terminal, but that's also probably the desired way there. Otherwise when you resize the window which has many tabs, each with a giant scrollback, the amount of time will be multiplied by the number of tabs.)

**Comment by monsta** [comment]
Hmm... are there any plans to add an async version of that function in VTE?

**Comment by egmontkob** [comment]
Nope, not at all. It would require way too much work for such a low benefit, and would have other downsides. We've put quite a lot of effort in finding the best compromise and implement that.

This is the exact sole reason why the API was introduced to disable rewrapping. It's a slow operation, hence the user should be able to disable it.

**Comment by monsta** [comment]
Then if we add this option, maybe it should be disabled by default?

@flexiondotorg @raveit65 what do you think about this feature guys?

**Comment by egmontkob** [comment]
This is the coolest feature vte/g-t has received in the last few years, and every feedback I've seen was about super happy users. Please don't disable rewrapping by default. Noticeable regression only starts occurring around 1M lines of scrollback which is way above the default and the typical. If you're concerned, you can show a warning to the user if rewrapping is enabled and the scrollback size is above a threshold (let's say 1M).

**Comment by monsta** [comment]
Ok, we need to implement and test this first anyway, then decide :smile:

**Comment by marcus-lundgren** [comment]
This is a feature that I really miss as I try to see if the MATE DE is a good replacement for Cinnamon DE. If you're worried about performance, then have it default to disabled. Let me as a user make an active decision, which I then can actively disable if needed. I think that's a good middle ground for us that come from other DEs, expecting this feature to exist.

**Comment by marcus-lundgren** [comment]
Hmm. Isn't this actually implemented? As far as I can see, this is defaulted to TRUE[1]. Since mate-terminal doesn't seem to set this to FALSE, then this is already being used! The text seems to reflow on Ubuntu MATE 18.04.

[1] https://developer.gnome.org/vte/unstable/VteTerminal.html#VteTerminal--rewrap-on-resize
